core/src/storage/collections/binary_heap/impls.rs

```diff
@@ -0,0 +1,461 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::storage::{
+    self,
+    alloc::{
+        Allocate,
+        AllocateUsing,
+        Initialize,
+    },
+    chunk::SyncChunk,
+    Flush,
+};
+use core::cmp::{
+    Ord,
+    Ordering
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+use super::access_wrapper::AccessWrapper;
+
+/// We implement a binary tree.
+pub const CHILDREN: u32 = 2;
+
+/// A binary heap collection.
+/// The heap depends on `Ord` and is a max-heap by default. In order to
+/// make it a min-heap implement the `Ord` trait explicitly on the type
+/// which is stored in the heap.
+///
+/// Provides `O(log(n))` push and pop operations.
+#[derive(Debug)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct BinaryHeap<T> {
+    /// Stores densely packed general heap information.
+    header: storage::Value<BinaryHeapHeader>,
```

Comment:
Do we need the header here for just one element or is it expected to have more header elements in the future?

Reply:
Good point, I removed it now and replaced it with `len` (I don't expect more fields).